src/core/overflowsafe_type.hpp

```diff
@@ -134,72 +150,81 @@ class OverflowSafeInt
-	inline OverflowSafeInt operator * (const int64  factor) const { OverflowSafeInt result = *this; result *= factor; return result; }
-	inline OverflowSafeInt operator * (const int    factor) const { OverflowSafeInt result = *this; result *= (int64)factor; return result; }
-	inline OverflowSafeInt operator * (const uint   factor) const { OverflowSafeInt result = *this; result *= (int64)factor; return result; }
-	inline OverflowSafeInt operator * (const uint16 factor) const { OverflowSafeInt result = *this; result *= (int64)factor; return result; }
-	inline OverflowSafeInt operator * (const byte   factor) const { OverflowSafeInt result = *this; result *= (int64)factor; return result; }
+	inline OverflowSafeInt  operator *  (const OverflowSafeInt& factor) const { OverflowSafeInt result = *this; result *= factor; return result; }
+	template <class I, class = typename std::enable_if<std::is_integral<I>::value>::type>
+	inline OverflowSafeInt& operator *= (const I                factor)       { return *this *= OverflowSafeInt(factor); }
+	template <class I, class = typename std::enable_if<std::is_integral<I>::value>::type>
+	inline OverflowSafeInt  operator *  (const I                factor) const { return *this *  OverflowSafeInt(factor); }
 
 	/* Operators for division */
-	inline OverflowSafeInt& operator /= (const int64            divisor)       { this->m_value /= divisor; return *this; }
-	inline OverflowSafeInt  operator /  (const OverflowSafeInt& divisor) const { OverflowSafeInt result = *this; result /= divisor.m_value; return result; }
-	inline OverflowSafeInt  operator /  (const int              divisor) const { OverflowSafeInt result = *this; result /= divisor; return result; }
-	inline OverflowSafeInt  operator /  (const uint             divisor) const { OverflowSafeInt result = *this; result /= (int)divisor; return result; }
+	template <class I, class = typename std::enable_if<std::is_integral<I>::value && (static_cast<intmax_t>(std::numeric_limits<I>::min()) >= static_cast<intmax_t>(T_MIN)) && (static_cast<uintmax_t>(std::numeric_limits<I>::max()) <= static_cast<uintmax_t>(T_MAX))>::type>
+	inline OverflowSafeInt& operator /= (const I                divisor)       { this->m_value /= static_cast<T>(divisor); return *this; }
+	template <class I, class = typename std::enable_if<std::is_integral<I>::value && (static_cast<intmax_t>(std::numeric_limits<I>::min()) >= static_cast<intmax_t>(T_MIN)) && (static_cast<uintmax_t>(std::numeric_limits<I>::max()) <= static_cast<uintmax_t>(T_MAX))>::type>
+	inline OverflowSafeInt  operator /  (const I                divisor) const { OverflowSafeInt result = *this; result /= static_cast<T>(divisor); return result; }
+	inline OverflowSafeInt& operator /= (const intmax_t         divisor)       { *this = *this / divisor; return *this; }
+	inline OverflowSafeInt  operator /  (const intmax_t         divisor) const { return OverflowSafeInt(static_cast<intmax_t>(this->m_value) / static_cast<intmax_t>(divisor)); }
```

Comment:
I am not 100% sure this construct will work as intended. The code does at least compile, though.